File: restart_server/restart_server.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <unistd.h>

int main(int argc, char **argv) {

    char buffer[75];

    sprintf(buffer, "/usr/local/bin/restart_server.py %s", argv[1]);

    printf("Running command: %s\n\n", buffer);

    setuid(0);
    system(buffer);

    return 0;
}

File: restart_server/restart_server.py
#!/usr/bin/python

import sys

from fabric.api import local


def restart_nginx():
    local("nginx -t")
    local("/etc/init.d/nginx restart")


def restart_apache():
    local("apache2ctl -t")
    local("/etc/init.d/apache2 restart")


def restart_uwsgi(version):
    local("supervisorctl restart %s" % (version,))


if __name__ == '__main__':
    command = sys.argv[1]

    if command.startswith("uwsgi"):
        command = command.replace('.', '')
        restart_uwsgi(command)
    else:
        locals()['restart_%s' % (command,)]()
